core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/ChannelMarkerManager.scala

```diff
@@ -56,35 +63,61 @@ class ChannelMarkerManager(val actorId: ActorVirtualIdentity, inputGateway: Inpu
       marker: ChannelMarkerPayload
   ): Boolean = {
     val markerId = marker.id
+    val portId = inputGateway.getChannel(from).getPortId
     if (!markerReceived.contains(markerId)) {
-      markerReceived(markerId) = Set()
+      markerReceived(markerId) = new mutable.HashMap[PortIdentity, Set[ChannelIdentity]]()
     }
-    markerReceived.update(markerId, markerReceived(markerId) + from)
+    val portMap = markerReceived(markerId)
+    if (!portMap.contains(portId)) {
+      portMap(portId) = Set()
+    }
+    portMap.update(portId, portMap(portId) + from)
     // check if the epoch marker is completed
-    val markerReceivedFromAllChannels =
-      getChannelsWithinScope(marker).subsetOf(markerReceived(markerId))
+
     val epochMarkerCompleted = marker.markerType match {
-      case REQUIRE_ALIGNMENT => markerReceivedFromAllChannels
-      case NO_ALIGNMENT      => markerReceived(markerId).size == 1 // only the first marker triggers
+      case REQUIRE_ALIGNMENT =>
+        val markerReceivedFromAllChannels =
+          getChannelsWithinScope(marker).subsetOf(portMap.values.flatten.toSet)
+        if (markerReceivedFromAllChannels) {
+          markerReceived.remove(markerId) // clean up if all markers are received
+        }
+        markerReceivedFromAllChannels
+      case PORT_ALIGNMENT =>
+        val markerReceivedFromCurrentPort =
+          getChannelsWithinPort(marker, portId).subsetOf(portMap(portId))
+        if (markerReceivedFromCurrentPort) {
+          portMap.remove(portId)
+        }
+        markerReceivedFromCurrentPort
+      case NO_ALIGNMENT =>
+        markerReceived(markerId).size == 1 // only the first marker triggers
```

Comment:
this is not correct because the marker from the same port will trigger it again.

Reply:
The change of markerReceived is reverted.